lib/ProductOpener/Ingredients.pm

```diff
@@ -2332,16 +2332,54 @@ sub compute_ingredients_percent_values($$$) {
 	return $changed_total;
 }
 
+
+=head2 init_percent_values($total_min, $total_max, $ingredients_ref)
+
+Initialize the percent, percent_min and percent_max value of a list of ingredients.
+
+When a percent is specifically set, use this value for percent_min and percent_max.
+
+Warning: percent listed for sub-ingredients can be absolute (e.g. "Sugar, fruits 40% (pear 30%, apple 10%)")
+or they can be relative to the parent ingredient (e.g. "Sugar, fruits 40% (pear 75%, apple 25%)".
+
+Otherwise use 0 for percent_min and total_max for percent_max.
+
+=cut
+
 sub init_percent_values($$$) {
 
 	my $total_min = shift;
 	my $total_max = shift;
 	my $ingredients_ref = shift;
 
+	# Determine if percent listed are absolute (default) or relative to a parent ingredient
+
+	my $percent_mode = "absolute";
+
+	# Assume that percent listed can be relative only if there is a specified percent for the parent ingredient
+	# in which case $total_min is equal to $total_max and less than 100.
```

Comment:
Actually this exact example is covered because even though we don't have a % for fruit concentrate, since there are only 2 parents ingredients, we determine that percent_min and percent_max is 40%.

So we get this result:

`[
  {
    'id' => 'en:water',
    'percent' => 60,
    'percent_estimate' => 60,
    'percent_max' => 60,
    'percent_min' => 60,
    'text' => 'water'
  },
  {
    'id' => 'en:fruit-concentrates',
    'ingredients' => [
      {
        'id' => 'en:apple',
        'percent' => 16,
        'percent_estimate' => 16,
        'percent_max' => 16,
        'percent_min' => 16,
        'text' => 'apple'
      },
      {
        'id' => 'en:mango',
        'percent' => 12,
        'percent_estimate' => 12,
        'percent_max' => 12,
        'percent_min' => 12,
        'text' => 'mango'
      },
      {
        'id' => 'en:citrus-fruit',
        'percent_estimate' => 12,
        'percent_max' => 12,
        'percent_min' => 12,
        'text' => 'citrus'
      }
    ],
    'percent_estimate' => 40,
    'percent_max' => 40,
    'percent_min' => 40,
    'text' => 'fruit concentrate'
  }
]
`

Reply:
What's more tricky is if we have percent_min different from percent_max, e.g. if we add a 3rd parent ingredient without a %.

water (60%), fruit concentrate (apple 40%, mango 30%, citrus), sugar

Then in this case, we can determine that those are relative %, but they are relative to a range, so we can't derive an exact percent. We could apply it to the min and max of apple and mango. I will try.